miio/push_server.py

```diff
@@ -0,0 +1,401 @@
+            self.unregister_miio_device(self._registered_devices[ip]["device"])
+
+        self._listen_couroutine.close()
+        self._listen_couroutine = None
+
+    def subscribe_event(self, device: Device, event_info: EventInfo):
+        """Subscribe to a event such that the device will start pushing data for that
+        event."""
+        if device.ip not in self._registered_devices:
+            _LOGGER.error("Can not subscribe event, miio device not yet registered")
+            return None
+
+        if self.server_ip is None:
+            _LOGGER.error("Can not subscribe event withouth starting the push server")
+            return None
+
+        self._event_id = self._event_id + 1
+        event_id = f"x.scene.{self._event_id}"
+
+        event_payload = self._construct_event(event_id, event_info, device)
+
+        response = device.send(
+            "send_data_frame",
+            {
+                "cur": 0,
+                "data": event_payload,
+                "data_tkn": 29576,
+                "total": 1,
+                "type": "scene",
+            },
+        )
+
+        if response != ["ok"]:
+            _LOGGER.error(
+                "Error subscribing event, response %s, event_payload %s",
+                response,
+                event_payload,
+            )
+            return None
+
+        event_ids = self._registered_devices[device.ip]["event_ids"]
+        event_ids.append(event_id)
+
+        return event_id
+
+    def unsubscribe_event(self, device: Device, event_id):
+        """Unsubscribe from a event by id."""
+        result = device.send("miIO.xdel", [event_id])
+        if result == ["ok"]:
+            event_ids = self._registered_devices[device.ip]["event_ids"]
+            if event_id in event_ids:
+                event_ids.remove(event_id)
+        else:
+            _LOGGER.error("Error removing event_id %s: %s", event_id, result)
+
+        return result
+
+    def _construct_event(  # nosec
+        self,
+        event_id,
+        info: EventInfo,
+        device: Device,
+    ):
+        """Construct the event data payload needed to subscribe to an event."""
+        if info.event is None:
+            info.event = info.action
+        if info.source_sid is None:
+            info.source_sid = str(device.device_id)
+        if info.source_model is None:
+            info.source_model = device.model
+        if info.source_token is None:
+            info.source_token = device.token
+
+        token_enc = calculated_token_enc(info.source_token)
+        source_id = info.source_sid.replace(".", "_")
+        command = f"{self.server_model}.{info.action}:{source_id}"
+        key = f"event.{info.source_model}.{info.event}"
+        message_id = 0
+
+        if len(command) > 49:
+            _LOGGER.error(
+                "push server event command can be max 49 chars long, '%s' is %i chars",
+                command,
+                len(command),
+            )
```

Comment:
If the lenght is too long you will still receive the callback message but the action received will be truncated at 49 chars, this will cause the device_id (which is at the end) to be (partially) cut off and therefore the source device of the callback can not be matched anymore.
So not too big of a deal but it will cause some problems, I don't think an execption is needed.

Reply:
I will clearify the error message

Reply:
What type of problems? If it makes it break the callback matching (i.e., expected functionality) and the issue can be fixed by the developer, it would be better to raise an exception to make it clear that something is not correct.

Reply:
Well yes it might break callback matching, I am fine to change it to a exception, I will change it.

Reply:
The only problem with a exception is that it will break then so no callbacks at all will work, while if you have a log error, the rest of the events will still work properly and only the one which is to long will not match and generate errors that it could not be matched.